src/cuda/encode_ptx.h

```diff
@@ -0,0 +1,449 @@
+	out[2] = x[2] ^ y[2];
+	out[3] = x[3] ^ y[3];
+	out[4] = x[4] ^ y[4];
+	out[5] = x[5] ^ y[5];
+	out[6] = x[6] ^ y[6];
+	out[7] = x[7] ^ y[7];
+}
+
+__device__ __forceinline__ void eq_x_y(u256 x, const u256 y)
+{  // x = y
+	x[0] = y[0];
+	x[1] = y[1];
+	x[2] = y[2];
+	x[3] = y[3];
+	x[4] = y[4];
+	x[5] = y[5];
+	x[6] = y[6];
+	x[7] = y[7];
+}
+
+__device__ __forceinline__ void addition_chain_reduce_256(u256 out, const u256 x)
+{
+	u256 temp;
+
+	square_mul_reduce_256(out, x, 1, x);
+	square_mul_reduce_256(temp, out, 1, x);
+	square_mul_reduce_256(out, temp, 3, temp);
+	square_mul_reduce_256(out, out, 1, x);
+	square_mul_reduce_256(out, out, 7, out);
+	square_mul_reduce_256(out, out, 14, out);
+	square_mul_reduce_256(out, out, 3, temp);
+	square_mul_reduce_256(out, out, 31, out);
+	square_mul_reduce_256(out, out, 62, out);
+	square_mul_reduce_256(out, out, 124, out);
+	square_mul_reduce_256(out, out, 2, x);
+	square_mul_reduce_256(out, out, 4, x);
+
+	if (check_ge_prime(out))
+		x_minus_prime(out);
+}
+
+__device__ __forceinline__ void sqrt_permutation_ptx(u256 out, u256 x)
+{   // finds the square root of x in SLOTH setting
+
+	addition_chain_reduce_256(out, x);  // out = pow(x, (prime+1)/4, prime)
+
+	if (check_odd(out)) {  // if the number is odd, we have to negate the square-root
+		prime_minus_x(out);
+	}
+
+	// check if square of x is the same with input (this check is necessary,
+	// since there are some numbers that do not have square roots,
+	// but their negations do have, read the SLOTH paper for more details).
+	// Then negate the result, and assign this to variable `neg`.
+	// So, if our computed square_root's square equals to input, this means
+	// the input indeed has a square root and we found it
+	u256 check_candidate;
+	mul_reduce_256(check_candidate, out, out);
+
+	// because of lazy-reduction in mul_reduce, we have to check the result is greater than prime
+	if (check_ge_prime(check_candidate)) {
+		x_minus_prime(check_candidate);  // if so, apply the reduction. One check is enough, lazy-reduction guarantees that
+	}
+
+
+	if (check_eq_x_y(check_candidate, x)) {}  // means we have found the correct square root
+	else {
+		if (check_even(out)) {
+			prime_minus_x(out);  // negate the square-root accordingly with it being even-odd.
+		}
+	}
+}
+
+__global__ void encode_ptx(u256* piece, u256* nonce, u256* farmer_id)
+{    // in this version, expanded_iv will be computed from nonce array and farmer_id
+
+	int global_idx = threadIdx.x + blockIdx.x * blockDim.x;  // global index of the thread
+
+	u256 feedback;
+	xor_x_y(feedback, nonce[global_idx], *farmer_id);  // computation of the expanded_iv from nonce array
+
+	for (int i = 0; i < 128; i++)  // actual sloth_encoding
+	{
+		xor_x_y(feedback, piece[i + global_idx * 128], feedback);
+		sqrt_permutation_ptx(piece[i + global_idx * 128], feedback);
+		eq_x_y(feedback, piece[i + global_idx * 128]);
+	}
+}
+
+__global__ void sloth256_189_encode(u256* piece_array, u256* expanded_iv)
```

Comment:
Wouldn't this conflict with CPU?

Reply:
I don't think so, but better to be safe, renaming